src/oop/object_interactions/medium/q4/MainTest.java

```diff
@@ -0,0 +1,220 @@
+        Object[][] test = new Object[][]{
+                {'A', 10.3, 'B', 4.1},
+                {'B', 4.2, 'A', 2.3},
+                {'O', 3.5, 'B', 5.5},
+        };
+        for (int num = 0; num < test.length; num++) {
+            char type = (char) test[num][0];
+            double volume = (double) test[num][1];
+            Object[][] arguments = {
+                    {type, char.class},
+                    {volume, double.class}
+            };
+            Object classInstance = testClass.createInstance(arguments);
+            _assertEquals(type, testClass.getFieldValue(classInstance, bloodTypeAttributeName),
+                    "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+            _assertEquals(volume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+            char updateType = (char) test[num][2];
+            double updateVolume = (double) test[num][3];
+            Object output = testClass.callMethod(getBloodTypeMethodName, classInstance);
+            _assertEquals(type, output,
+                    "You have modified the provided " + getBloodTypeMethodName + " method on the " + className + " class. Please revert it to the original state.");
+            output = testClass.callMethod(getVolumeMethodName, classInstance);
+            _assertEquals(volume, output,
+                    "You have modified the provided " + getVolumeMethodName + " method on the " + className + " class. Please revert it to the original state.");
+            output = testClass.callMethod(setBloodTypeMethodName, new Object[][]{{updateType, char.class}}, classInstance);
+            _assertEquals(updateType, testClass.getFieldValue(classInstance, bloodTypeAttributeName),
+                    "You have modified the provided " + setBloodTypeMethodName + " method on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setBloodTypeMethodName, "should not return any output"));
+            output = testClass.callMethod(setVolumeMethodName, new Object[][]{{updateVolume, double.class}}, classInstance);
+            _assertEquals(updateVolume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided " + setVolumeMethodName + " method on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setVolumeMethodName, "should not return any output"));
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
```

Comment:
You should not always use 10.3 and A here, the test cases should be more varied.

Reply:
Also, please add another test case where the recipient has less blood than the donor but the transfusion occurs.